tests/cli.rs

```diff
@@ -181,3 +181,21 @@ stylua = { github = "JohnnyMorganz/StyLua", version = "0.11.3" }
     );
     context.snapshot_command("install_invalid_auth_configuration");
 }
+
+#[test]
+fn snapshot_install_all_tools_before_failing() {
+    let mut context = TestContext::foreman().arg("install");
+    let config_path = context.path_from_working_directory("foreman.toml");
+    write_file(
+        &config_path,
+        r#"
+[tools]
+stylua = { github = "JohnnyMorganz/StyLua", version = "0.11.3" }
+not-a-real-tool = { github = "Roblox/VeryFakeRepository", version = "0.1.0" }
+badly-formatted-tool = { github = "Roblox/", version = "0.2.0" }
+selene = { source = "Kampfkarren/selene", version = "=0.22.0" }
```

Comment:
This test ends up actually downloading these tools. Is there any way to mock this data?

Reply:
This is possible yes but you will have to refactor the code to inject a custom ToolProvider implementation. This is probably easier to do within the `src` code. Tests in `tests` are considered at least integration level and so they cannot access private members. So if you refactor the code a little, you can write a unit test in `main.rs` that injects a memory only tool provider

Reply:
I think I'll just add an e2e test with the actual downloads and just test out the download failure here. I don't want the scope of this PR to get too large, so I might just file an issue to inject a mocked tool provider if there isn't one already.